module/os/linux/spl/spl-taskq.c

```diff
@@ -1255,12 +1283,105 @@ module_param_call(spl_taskq_kick, param_set_taskq_kick, param_get_uint,
 MODULE_PARM_DESC(spl_taskq_kick,
 	"Write nonzero to kick stuck taskqs to spawn more threads");
 
+#ifdef HAVE_CPU_HOTPLUG
+/*
+ * This callback will be called exactly once for each core that comes online,
+ * for each dynamic taskq.  There are two kinds of dynamic taskqs that we
+ * attempt to expand; those who have TASKQ_THREADS_CPU_PCT set, and those
+ * created with the same number of threads as we have cores. In the CPU_PCT
+ * case, we need to redo the percentage calculation each time this is
+ * called. For the threads == cores case, we always want to create a new
+ * thread to keep the number of cores and threads in sync.
+ */
+static int
+spl_taskq_expand(unsigned int cpu, struct hlist_node *node)
+{
+	taskq_t *tq = list_entry(node, taskq_t, tq_hp_cb_node);
+	unsigned long flags;
+	int err = 0;
+
+	ASSERT(tq);
+	spin_lock_irqsave_nested(&tq->tq_lock, flags, tq->tq_lock_class);
+
+	if (!(tq->tq_flags & TASKQ_ACTIVE))
+		goto out;
+
+	if (tq->tq_flags & TASKQ_THREADS_CPU_PCT) {
+		int nthreads = MIN(tq->tq_cpu_pct, 100);
+		nthreads = MAX(((num_online_cpus() + 1) * nthreads) / 100, 1);
+		tq->tq_maxthreads = nthreads;
+	} else {
+		tq->tq_maxthreads++;
+	}
+
+	if (!((tq->tq_flags & TASKQ_DYNAMIC) && spl_taskq_thread_dynamic) &&
+	    tq->tq_maxthreads > tq->tq_nthreads) {
+		taskq_thread_t *tqt = taskq_thread_create(tq);
+		if (tqt == NULL)
+			err = -1;
+	}
+
+out:
+	spin_unlock_irqrestore(&tq->tq_lock, flags);
+	return (err);
+}
+
+/*
+ * While we don't support offlining CPUs, it is possible that CPUs will fail
+ * to online successfully. We do need to be able to handle this case
+ * gracefully.
+ */
+static int
+spl_taskq_prepare_down(unsigned int cpu, struct hlist_node *node)
+{
+	taskq_t *tq = list_entry(node, taskq_t, tq_hp_cb_node);
+	unsigned long flags;
+	int err = 0;
+
+	ASSERT(tq);
+	spin_lock_irqsave_nested(&tq->tq_lock, flags, tq->tq_lock_class);
+
+	if (!(tq->tq_flags & TASKQ_ACTIVE))
+		goto out;
+
+	if (tq->tq_flags & TASKQ_THREADS_CPU_PCT) {
+		int nthreads = MIN(tq->tq_cpu_pct, 100);
+		nthreads = MAX(((num_online_cpus()) * nthreads) / 100, 1);
+		tq->tq_maxthreads = nthreads;
+	} else {
+		tq->tq_maxthreads--;
+	}
+
+	if (!((tq->tq_flags & TASKQ_DYNAMIC) && spl_taskq_thread_dynamic) &&
+	    tq->tq_maxthreads > tq->tq_nthreads) {
+		taskq_thread_t *tqt = list_entry(tq->tq_thread_list.next,
+		    taskq_thread_t, tqt_thread_list);
+		struct task_struct *thread = tqt->tqt_thread;
+		spin_unlock_irqrestore(&tq->tq_lock, flags);
+
+		kthread_stop(thread);
+
+		spin_lock_irqsave_nested(&tq->tq_lock, flags,
+		    tq->tq_lock_class);
```

Comment:
It looks like you could simply `return 0;` here and avoid needing to acquire and drop the lock again.  This might be nice since it would ensure that if somehow this was the last thread and the taskq was destroyed while the lock was dropped we wouldn't panic due to an invalid `tq` pointer.  While highly unlikely it does seem like this might be possible so I wanted to mention it.